src/modules/span_modules.py

```diff
@@ -0,0 +1,213 @@
+                predict: bool) -> Dict:
+        """ 
+        Run forward pass.
+        Expects batch to have the following entries:
+            'input' : [batch_size, max_len, emb_size]
+            'labels' : [batch_size, num_targets] of label indices
+            'span1s' : [batch_size, 1, 2], span indices
+            'span2s' : [batch_size, 1, 2], span indices 
+                . 
+                . 
+                . 
+            'span_ts': [batch_size, 1, 2], span indices
+
+        Parameters
+        -------------------------------
+            batch: dict(str -> Tensor) with entries described above.
+            sent_embs: [batch_size, max_len, repr_dim] Tensor
+            sent_mask: [batch_size, max_len, 1] Tensor of {0,1}
+            task: Task
+            predict: whether or not to generate predictions
+        This learns different span pooling operators for each span.
+        
+        Returns
+        -------------------------------
+            out: dict(str -> Tensor)
+        """
+        out = {}
+        cuda_device = -1
+        if torch.cuda.is_available():
+            cuda_device = torch.cuda.current_device()
+        batch_size = sent_embs.shape[0]
+        out["n_inputs"] = batch_size
+        # Apply projection CNN layer for each span of the input sentence 
+        sent_embs_t = sent_embs.transpose(1, 2)
+        sent_embs_t = move_to_device(sent_embs_t, cuda_device)
+        se_projs = []
+        for i in range(self.num_spans):
+            se_proj = self.projs[i](sent_embs_t).transpose(2, 1).contiguous()
+            se_projs.append(se_proj)
+
+        span_embs = torch.Tensor([]).cuda() \
+            if torch.cuda.is_available() else torch.Tensor([])
+        out["n_exs"] = batch_size
+        _kw = dict(sequence_mask=sent_mask.long())
+        for i in range(self.num_spans):
+            # spans are [batch_size, num_targets, span_modules]
+            span_emb = self.span_extractors[i](se_projs[i], batch["span" + str(i + 1) + "s"], **_kw)
+            span_embs = torch.cat([span_embs, span_emb], dim=2)
+
+        # [batch_size, num_targets, n_classes]
+        logits = self.classifier(span_embs)
+        out["logits"] = logits
+
+        # Compute loss if requested.
+        if "labels" in batch:
+            logits = logits.squeeze(dim=1)
+            out["loss"] = self.compute_loss(logits,
+                                            batch["labels"].squeeze(dim=1),
+                                            task)
+            predictions = self.get_predictions(logits)
+            tagmask = batch.get("tagmask", None)
+            task.update_metrics(predictions,  batch["labels"].squeeze(dim=1), tagmask=tagmask)
+
+        if predict:
+            # Return preds as a list.
+            preds = self.get_predictions(logits)
+            out["preds"] = list(self.unbind_predictions(preds))
+        return out
+
+    def unbind_predictions(self, preds: torch.Tensor) -> Iterable[np.ndarray]:
```

Comment:
seems more appropriate as a utils function than as something specific to this module

Reply:
delete since this is not in utils?